File: src/lib/desktop/fiscal.ts
import { invoke } from '@tauri-apps/api/core';
import type { TauriActionResult } from '@/lib/desktop/admin-types';
import { isTauriBackendAvailable } from '@/lib/desktop/runtime';

function requireTauri(): void {
  if (!isTauriBackendAvailable()) {
    throw new Error('Tauri fiscal API requires the desktop shell');
  }
}

export type TauriFiscalFlexibleResult = {
  success: boolean;
  data?: unknown;
  error?: string;
  skipped?: boolean;
  warning?: string;
};

export async function tauriFiscalGetSettings() {
  requireTauri();
  return invoke<TauriActionResult<{ settings: unknown; state: unknown }>>('fiscal_get_settings_cmd');
}

export async function tauriFiscalUpdateSettings(patch: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<void>>('fiscal_update_settings_cmd', { patch });
}

export async function tauriFiscalSyncConfig() {
  requireTauri();
  return invoke<TauriActionResult<unknown>>('fiscal_sync_config_cmd');
}

export async function tauriFiscalVerifyTaxpayer(
  deviceId: number,
  activationKey: string,
  environment: 'test' | 'production'
) {
  requireTauri();
  return invoke<TauriActionResult<unknown>>('fiscal_verify_taxpayer_cmd', {
    deviceId,
    activationKey,
    environment,
  });
}

export async function tauriFiscalPingDevice() {
  requireTauri();
  return invoke<TauriActionResult<unknown>>('fiscal_ping_device_cmd');
}

export async function tauriFiscalRefreshStatus() {
  requireTauri();
  return invoke<TauriActionResult<unknown>>('fiscal_refresh_status_cmd');
}

export async function tauriFiscalOpenDay() {
  requireTauri();
  return invoke<TauriActionResult<unknown>>('fiscal_open_day_cmd');
}

export async function tauriFiscalCloseDay() {
  requireTauri();
  return invoke<TauriActionResult<unknown>>('fiscal_close_day_cmd');
}

export async function tauriFiscalRegisterDevice(input: {
  environment: 'test' | 'production';
  deviceId: number;
  deviceSerialNo: string;
  activationKey: string;
  deviceModelName?: string;
  deviceModelVersion?: string;
}) {
  requireTauri();
  return invoke<TauriActionResult<unknown>>('fiscal_register_device_cmd', {
    environment: input.environment,
    deviceId: input.deviceId,
    deviceSerialNo: input.deviceSerialNo,
    activationKey: input.activationKey,
    deviceModelName: input.deviceModelName ?? null,
    deviceModelVersion: input.deviceModelVersion ?? null,
  });
}

export async function tauriFiscalisePosSale(input: {
  orderId?: string;
  invoiceNo: string;
  items: Array<{ name: string; price: number; quantity: number; taxPercent?: number }>;
  payments: Array<{ method: string; amount: number }>;
  taxEnabled?: boolean;
}) {
  requireTauri();
  return invoke<TauriFiscalFlexibleResult>('fiscal_fiscalise_sale_cmd', {
    orderId: input.orderId ?? null,
    invoiceNo: input.invoiceNo,
    items: input.items,
    payments: input.payments,
    taxEnabled: input.taxEnabled ?? null,
  });
}

export async function tauriFiscalisePosCreditNote(input: {
  orderId: string;
  receiptNotes?: string;
}) {
  requireTauri();
  return invoke<TauriFiscalFlexibleResult>('fiscal_fiscalise_credit_note_cmd', {
    orderId: input.orderId,
    receiptNotes: input.receiptNotes ?? null,
  });
}

File: src-tauri/src/commands/fiscal.rs
use serde_json::Value;
use tauri::State;

use crate::admin::service::ActionResult;
use crate::auth::session::SessionState;
use crate::fiscal::service::{
    close_fiscal_day, fiscalise_credit_note, fiscalise_sale, get_settings, open_fiscal_day,
    ping_device, refresh_status, register_device, sync_config, update_settings, verify_taxpayer,
    FiscalFlexibleResult,
};
use crate::state::AppState;

#[tauri::command]
pub fn fiscal_get_settings_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
) -> ActionResult<Value> {
    get_settings(&app, &session)
}

#[tauri::command]
pub fn fiscal_update_settings_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    patch: Value,
) -> ActionResult<()> {
    update_settings(&app, &session, patch)
}

#[tauri::command]
pub fn fiscal_sync_config_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
) -> ActionResult<Value> {
    sync_config(&app, &session)
}

#[tauri::command]
pub fn fiscal_verify_taxpayer_cmd(
    device_id: i64,
    activation_key: String,
    environment: String,
) -> ActionResult<Value> {
    verify_taxpayer(device_id, activation_key, environment)
}

#[tauri::command]
pub fn fiscal_ping_device_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
) -> ActionResult<Value> {
    ping_device(&app, &session)
}

#[tauri::command]
pub fn fiscal_refresh_status_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
) -> ActionResult<Value> {
    refresh_status(&app, &session)
}

#[tauri::command]
pub fn fiscal_open_day_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
) -> ActionResult<Value> {
    open_fiscal_day(&app, &session)
}

#[tauri::command]
pub fn fiscal_close_day_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
) -> ActionResult<Value> {
    close_fiscal_day(&app, &session)
}

#[tauri::command]
pub fn fiscal_register_device_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    environment: String,
    device_id: i64,
    device_serial_no: String,
    activation_key: String,
    device_model_name: Option<String>,
    device_model_version: Option<String>,
) -> ActionResult<Value> {
    register_device(
        &app,
        &session,
        environment,
        device_id,
        device_serial_no,
        activation_key,
        device_model_name,
        device_model_version,
    )
}

#[tauri::command]
pub fn fiscal_fiscalise_sale_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    order_id: Option<String>,
    invoice_no: String,
    items: Vec<Value>,
    payments: Vec<Value>,
    tax_enabled: Option<bool>,
) -> FiscalFlexibleResult {
    fiscalise_sale(
        &app,
        &session,
        order_id,
        invoice_no,
        items,
        payments,
        tax_enabled,
    )
}

#[tauri::command]
pub fn fiscal_fiscalise_credit_note_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    order_id: String,
    receipt_notes: Option<String>,
) -> FiscalFlexibleResult {
    fiscalise_credit_note(&app, &session, order_id, receipt_notes)
}
